runtime/src/blob.rs

```diff
@@ -0,0 +1,233 @@
+#[derive(Encode, Decode, Clone, PartialEq, Debug)]
+pub struct Blob {
+    id: BlobId,
+    blob: Vec<u8>,
+    author: did::Did,
+}
+
+pub trait Trait: system::Trait + did::Trait {
+    /// Blobs larger than this will not be accepted.
+    type MaxBlobSize: Get<u32>;
+}
+
+decl_error! {
+    /// Error for the token module.
+    pub enum BlobError for Module<T: Trait> {
+        /// The blob is greater than `MaxBlobSize`
+        BlobTooBig,
+        /// There is already a blob with same id
+        BlobAlreadyExists,
+        /// There is no such DID registered
+        DidDoesNotExist,
+        /// Signature verification failed while adding blob
+        InvalidSig
+    }
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Blob {
+        Blobs get(fn get_blob): map hasher(blake2_128_concat)
+            dock::blob::BlobId => Option<(dock::did::Did, Vec<u8>)>;
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        /// Create a new immutable blob.
+        #[weight = 10_000]
+        pub fn new(
+            origin,
+            blob: dock::blob::Blob,
+            signature: dock::did::DidSignature,
+        ) -> DispatchResult {
+            Module::<T>::new_(origin, blob, signature)
+        }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    fn new_(
+        origin: <T as system::Trait>::Origin,
+        blob: Blob,
+        signature: did::DidSignature,
+    ) -> DispatchResult {
+        ensure_signed(origin)?;
+
+        // check
+        ensure!(
+            T::MaxBlobSize::get() as usize >= blob.blob.len(),
+            BlobError::<T>::BlobTooBig
+        );
+        ensure!(
+            !Blobs::contains_key(&blob.id),
+            BlobError::<T>::BlobAlreadyExists
+        );
+        let payload = crate::StateChange::Blob(blob.clone()).encode();
+        let valid = did::Module::<T>::verify_sig_from_did(&signature, &payload, &blob.author)?;
+        ensure!(valid, BlobError::<T>::InvalidSig);
+
+        // execute
+        Blobs::insert(blob.id, (blob.author, blob.blob));
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_common::*;
+
+    fn rando_bytes(len: usize) -> Vec<u8> {
+        let ret: Vec<u8> = (0..len).map(|_| rand::random()).collect();
+        assert_eq!(ret.len(), len);
+        ret
+    }
+
+    fn create_blob(
+        id: BlobId,
+        content: Vec<u8>,
+        author: did::Did,
+        author_kp: sr25519::Pair,
+    ) -> DispatchResult {
+        let bl = Blob {
+            id,
+            blob: content,
+            author,
+        };
+        let sig = sign(&crate::StateChange::Blob(bl.clone()), &author_kp);
+        BlobMod::new(Origin::signed(ABBA), bl.clone(), sig)
+    }
+
+    fn get_max_blob_size() -> usize {
+        <Test as crate::blob::Trait>::MaxBlobSize::get() as usize
+    }
+
+    #[test]
+    fn add_blob() {
+        if !in_ext() {
```

Comment:
Same comment on the extra call to the test function as one of the earlier PRs. Let's address this with one of the approaches discussed there.

Reply:
Switched to alternate style in 99a5659120f7f1238a5f93831f3f9de376932e31:

```rust
ext().execute_with(|| {
    ...
});
```

Added new issue https://github.com/docknetwork/dock-substrate/issues/44